backend/dal/Services/Concrete/ProjectReportService.cs

```diff
@@ -0,0 +1,285 @@
+        /// <returns></returns>
+        public IEnumerable<ProjectSnapshot> Refresh(int reportId)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.ReportsSpl);
+            var report = Get(reportId);
+
+            var generatedSnapshots = GenerateSnapshots(DateTime.UtcNow, report.From);
+
+            return generatedSnapshots;
+        }
+
+        /// <summary>
+        /// Remove the specified report from the datasource.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'report' is required.</exception>
+        /// <exception cref="RowVersionMissingException">Project report rowversion is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project report does not exist.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to delete report.</exception>
+        /// <returns></returns>
+        public void Remove(ProjectReport report)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.ReportsSpl);
+            if (report?.Id == null) throw new ArgumentNullException();
+
+            var originalReport = this.Context.ProjectReports
+                .SingleOrDefault(p => p.Id == report.Id) ?? throw new KeyNotFoundException();
+            if (originalReport.IsFinal) throw new InvalidOperationException($"Unable to delete FINAL project reports.");
+
+            this.Context.ProjectReports.Remove(originalReport);
+            this.Context.CommitTransaction();
+        }
+
+
+        /// <summary>
+        /// Generate snapshots for all SPL projects using the passed 'to' date as the snapshot date.
+        /// </summary>
+        /// <param name="to"></param>
+        /// <returns></returns>
+        private IEnumerable<ProjectSnapshot> GenerateSnapshots(DateTime to, DateTime? from)
+        {
+            var splProjects = this.Context.Projects
+                .Include(p => p.Agency)
+                .Include(p => p.Status)
+                .Where(p => p.Workflow.Code == "SPL");
+
+            Dictionary<int?, ProjectSnapshot> fromSnapshots = new Dictionary<int?, ProjectSnapshot>();
+            if (from != null) {
+                fromSnapshots = this.Context.ProjectSnapshots
+                    .Include(s => s.Project).ThenInclude(p => p.Agency)
+                    .Include(s => s.Project).ThenInclude(p => p.Status)
+                    .Where(s => s.SnapshotOn == from)
+                    .ToDictionary(s => (int?)s.ProjectId, s => s);
+                if (fromSnapshots.Count() == 0) throw new KeyNotFoundException("If specified report FROM value must return valid snapshots");
+            }
+
+            IEnumerable<ProjectSnapshot> projectSnapshots = new HashSet<ProjectSnapshot>();
+            
+            foreach (Project project in splProjects)
+            {
+                JsonConvert.PopulateObject(project.Metadata ?? "{}", project);
```

Comment:
Not ideal, but until the backend is refactored to not require newtonsoft I still need to do this.